File: 10-RegularExpressionMatching/10-RegularExpressionMatching.py
# Last updated: 16/5/2025, 11:01:06 pm
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        # abasdsadsdcsacdeeeff a.*cdeeeff
        m = len(s)
        n = len(p)

        @cache
        def rec(left, right):
            if right == n: return left == m
            elif right+1 < n and p[right+1] == '*' and rec(left, right+1): return True
            elif left == m: return p[right] == '*' and rec(left, right+1)
            
            if s[left] == p[right]:
                return rec(left+1, right+1)
            elif p[right] == '.':
                return rec(left+1, right+1)
            elif p[right] == '*':
                if rec(left, right+1): return True
                elif p[right-1] == '.' or p[right-1] == s[left]:
                    return rec(left+1, right)

            return False

        return rec(0, 0)

File: 10-RegularExpressionMatching/10-RegularExpressionMatching.cpp
// Last updated: 12/6/2025, 5:55:33 am
class Solution {
public:
    bool isMatch(string s, string p) {
        if (p.length() == 0) return s.length() == 0;
        bool firstmatch = (s.length() > 0 && (s[0] == p[0] || p[0] == '.'));
        
        if (p.length() >=2 && p[1] == '*'){
            return (isMatch(s, p.substr(2)) || (firstmatch && isMatch(s.substr(1),p)));
        }else{
            return (firstmatch && isMatch(s.substr(1),p.substr(1)));
        }
    }
};